File: machine learning models/Decision Tree Bankrunptcy Prediction/Python/bankrupcy.py
import numpy as np
import math
import random

###### function to select a variable to split
def infogain(x, y):
	info_gains = np.zeros(np.size(x, 1))
	# calculate information i(N)
	classes = np.unique(y)  # only binary 0/1
	iN = 0
	for class_value in classes:
		py = list(y).count(class_value) / np.size(y, 0)
		iN = iN - py * math.log(py) / math.log(2)

	# iterate over all variables on columns
	for col in range(np.size(x, 1)):
		values = np.unique(x[:, col]) # only binary 0/1
		# entropy
		iNx = 0
		for class_value in values:
			pf = list(x[:, col]).count(class_value) / np.size(x, 0)
			yf = y[x[:, col]==class_value]
			classes = np.unique(yf) # only binary 0/1
			iNk = 0
			for class_value in classes:
				pyf = list(yf).count(class_value) / np.size(yf, 0)
				iNk = iNk - pyf * math.log(pyf) / math.log(2)
			iNx = iNx + pf * iNk
		info_gains[col] = iN - iNx

	max_gain_feature = np.argmax(info_gains)
	gain = np.amax(info_gains)
	return {'gain':gain, 'max_gain_feature':max_gain_feature}


##### function to find used variables
def varpath(tree, node):
	curdepth = tree[node]['depth']  # get the depth of the node
	varis = [dict() for i in range(curdepth)] # var is used in python
	vals = [dict() for i in range(curdepth)]
	while curdepth > 0:
		current = tree[node] # current node
		parent = tree[tree[node]['parent']] # parent node
		# which variable was the parent node split on?
		varis[curdepth-1] = parent['variable']
		# if the node is the right child, the value=1, if it is the left, the value = 0
		vals[curdepth-1] = parent['rightchild'] == node
		# move up to the parent
		node = current['parent']
		curdepth = curdepth - 1
	return {'varis': varis, 'vals': vals}


##### function to count distribution on leaf
def leafcat(tree, node, X, Y):
	# return the counts of classes and from there to classify instances	
	variables = varpath(tree, node)['varis']
	values = varpath(tree, node)['vals']
	mask = list()
	for i in range(np.size(X, 0)):
		if np.all(X[i][variables] == values):
			mask.append(i)
	num_y0 = list(Y[mask]).count(0)
	num_y1 = list(Y[mask]).count(1)
	return {'num_y0': num_y0, 'num_y1': num_y1}


##### function to print tree
def printree(tree):
	traverse(tree, 0, "")
	return()

def traverse(tree, node, s):
	if not(tree[node]['isLeaf']):
		if len(s) > 0:
			print(s)
		traverse(tree, tree[node]['leftchild'], s+"V"+str(tree[node]['variable'])+'=0 ')
		traverse(tree, tree[node]['rightchild'], s+"V"+str(tree[node]['variable'])+'=1 ')
	else:
		y0 = tree[node]['labels']['num_y0']
		y1 = tree[node]['labels']['num_y1']
		s = s + '=> classify y='+ str((y1>y0)*1) + ', distribution = ['+str(y0)+':'+str(y1)+']'
		print(s)
	return()


def buildtree(X, Y, maxdepth):
	global index
	if np.size(X, 1) < maxdepth:
		maxdepth = size(X, 1)
	tree = [dict() for i in range(2**(maxdepth+1))]
	index = 0 # change for python 
	curnode = 0 # change for python
	parent = -1
	tree = build(tree, parent, curnode, X, Y, maxdepth)
	return tree

def build(tree, parent, curnode, X, Y, maxdepth):
	global index
	if index == 0:
		tree[curnode] = {'parent': parent, 'depth': 0}
	else:
		tree[curnode] = {'parent': parent, 'depth': tree[parent]['depth']+1}
	isLeaf = not(tree[curnode]['depth'] < maxdepth)
	tree[curnode]['isLeaf'] = isLeaf
	if not(isLeaf):
		# choose split var
		used_vars = varpath(tree, curnode)['varis']
		availVars = np.delete(range(0, np.size(X, 1)), used_vars)
		# select a variable to split
		bestIndex = infogain(X[:, availVars], Y)['max_gain_feature']
		tree[curnode]['variable'] = availVars[bestIndex]
		# add child nodes
		leftchild = index + 1
		rightchild = index + 2
		index = index + 2
		tree[curnode]['leftchild'] = leftchild
		tree[curnode]['rightchild'] = rightchild
		# keep generating tree
		v = availVars[bestIndex]
		tree = build(tree, curnode, leftchild, X[X[:, v] == 0, :], Y[X[:, v] == 0], maxdepth)
		tree = build(tree, curnode, rightchild, X[X[:, v] == 1, :], Y[X[:, v] == 1], maxdepth)
	else:
		# count obs in each class at the leaf nodes
		tree[curnode]['labels'] = leafcat(tree, curnode, X, Y)
	return tree

def classify(tree, X):
	n = np.size(X, 0)
	pred = np.zeros(n)
	for i in range(n):
		node = 0
		while True:
			if not(tree[node]['isLeaf']):
				var = tree[node]['variable']
				if X[i, var] == 1:
					node = tree[node]['rightchild']
				else:
					node = tree[node]['leftchild']
			else:
				label = tree[node]['labels']
				if label['num_y0'] == label['num_y1']:
					pred[i] = random.choice([True, False]) #random choice if distribution is equal
				else:
					pred[i] = (label['num_y1']>label['num_y0'])*1
				break
	return(pred)

import pandas
import numpy as np
import math
import random

valnum = 2 # binary situation
train_pc = 0.7
depth = 4

# read data from xls file
datPath = "../data/bankruptcy.xls"

# load data, remove first row of names, skip first column
datafile = pandas.read_excel(datPath, skiprows = 0, parse_cols="B:N")
# convert to matrix
data = datafile.as_matrix()

x = np.array(data[:, 0:12])
y = np.array(data[:, 12]) # keep it simple for binary situation

for i in range(np.size(x, 1)):
	if len(np.unique(x[:, i]))>valnum: # if continuous need to discretize
		binEdges = np.linspace(np.amin(x[:, i]), np.amax(x[:, i]), num = valnum + 1)
		binEdges[valnum] = binEdges[valnum]+1
		binIdx = np.digitize(x[:, i], bins=binEdges)
		x[:, i] = binIdx - 1 #0/1 values instead of 1/2 values

m = np.size(x, 0)
n = np.size(x, 1)
#random.seed(503) # set random seed
size1 = math.floor(train_pc * sum(y == 0))
size2 = math.floor(train_pc * sum(y == 1))
indY1 = np.linspace(0, len(y) - 1, len(y), dtype="int")[y == 0]
indY2 = np.linspace(0, len(y) - 1, len(y), dtype="int")[y == 1]
train_ind = np.concatenate((np.random.choice(indY1, size1, False),
                          (np.random.choice(indY2, size2, False))), axis=0)
yTrain = y[train_ind]
yTest = np.delete(y, train_ind, axis=0)
xTrain = x[train_ind, :]
xTest = np.delete(x, train_ind, axis=0)

tree = buildtree(xTrain, yTrain, depth)
printree(tree)

trainpred = classify(tree, xTrain)
print("Train Accuracy: %f\n" % (np.mean(yTrain == trainpred) * 100))

testpred = classify(tree, xTest)
print("Test Accuracy: %f\n" % (np.mean(yTest == testpred) * 100))

File: machine learning models/Decision Tree Bankrunptcy Prediction/MATLAB/bankrupcy.m
function main()
	clear all;
	[data,txt,raw] = xlsread('bankruptcy.xls','Sheet1');
	valnum = 2; % binary 0/1, note this code only work for binary data, can extend
	train_pc = 0.7;
	depth = 4;

	x = data(:,1:12);
	y = (data(:,13) == 1);
	for i=1:size(x,2)
		if (length(unique(x(:,i)))>valnum) % if continuous need to discretize 
			binEdges = linspace(min(x(:,i)),max(x(:,i)),valnum+1); % create edges for discretization
			[~,binIdx] = histc(x(:,i),[binEdges(1:end-1) Inf]); % binIdx is the discretized values, which is actually bin number
			x(:,i) = binIdx-1; % values 0/1 instead of 1/2
		end
	end

	[m,n] = size(x);
	% stratified sampling
	train_ind = [randsample(find(y==0),floor(train_pc*sum(y==0)));randsample(find(y==1),floor(train_pc*sum(y==1)))];

	yTrain = y(train_ind,:);
	xTrain = x(train_ind,:);
	tree = buildtree(xTrain, yTrain, depth); % create a tree recursively
	tree = tree(~cellfun('isempty',tree)); % remove unnessary cells which are empty 

	print(tree);

	trainpred = classify(tree, xTrain);
	disp('Training Accuracy:');
	mean(trainpred==yTrain)

	yTest = y(setdiff((1:m),train_ind),:);
	xTest = x(~ismember(1:m,train_ind),:);
	testpred = classify(tree, xTest);
	disp('Test Accuracy:');
	mean(testpred==yTest)
	
	function [vars,vals] = varpath(tree, node)
		curdepth = tree{node}.depth; % get the depth of the node
		vars = zeros(tree{node}.depth, 1);
		vals = vars;
		while curdepth>0
		  current = tree{node}; % current node
		  parent = tree{tree{node}.parent}; % parent node
		  vars(curdepth) = parent.variable; % which variable was the parent node split on?
		  vals(curdepth) = parent.rightchild==node;   % if the node is the right child, the value=1, if it is the left, the value = 0
		  node = current.parent; % move up to the parent
		  curdepth = curdepth - 1; % and reduce the depth correspondingly
		end
	end
	
	function print(tree)
		% Depth First Search
		traverse(tree, 1, '');
	end

	function s = traverse(tree, node, s)
		if ~tree{node}.isLeaf
		  if (length(s) > 0)
			disp(s);
		  end
		  traverse(tree, tree{node}.leftchild, strcat(s,' V',num2str(tree{node}.variable),'=0'));
		  traverse(tree, tree{node}.rightchild, strcat(s,' V',num2str(tree{node}.variable),'=1'));
		else
		  s = strcat(s,' => classify y=', num2str(tree{node}.labels(2)>tree{node}.labels(1)), ' distribution=[',num2str(tree{node}.labels(1)),', ', num2str(tree{node}.labels(2)),']');
		  disp(s)
		end  
	end

	function [num_y0, num_y1] = leafcat(tree, node, X, Y)
		% return the counts of classes and from there to classify instances
		[variables,values] = varpath(tree, node); % get the variables from the node up to the root
		[n,~] = size(X);
		mask = true(n,1);
		for i=1:size(variables,1) % iterate through variables/columns
		  mask = mask & (X(:, variables(i)) == values(i)); % mark instances that were split on each node/variable. Finally mask would have instances onto the current leaf node
		end
		num_y1 = sum(Y(mask)); % only good for binary
		num_y0 = n - num_y1; % only good for binary
	end
	
	function [gain, max_gain_feature] = infogain(x,y)
		info_gains = zeros(1, size(x,2));

		% calculate information i(N)
		classes = unique(y)';
		iN = 0;
		for c=classes
			py = sum(y==c)/size(y,1);
			iN = iN - py*log2(py);
		end

		% iterate over all variables on columns
		for col=1:size(x,2)
			values = unique(x(:,col))';
			% entropy
			iNx = 0;
			for f=values % only binary 0/1 for now
				pf = sum(x(:,col)==f)/size(x,1);
				yf = y(find(x(:,col)==f));
				
				classes = unique(yf)';
				iNk = 0;
				for c=classes
					pyf = sum(yf==c)/size(yf,1);
					iNk = iNk - pyf*log2(pyf);
				end
				iNx = iNx + pf * iNk;
			end
			info_gains(col) = iN - iNx;
		end
		[gain, max_gain_feature] = max(info_gains);
	end
	
	function pred = classify(tree, X)
		[n,~] = size(X);
		pred = zeros(n,1);

		for i=1:n
			node = 1; % start with the root node
			while true % for each test instance, go through the trained tree until the leaf
				if ~tree{node}.isLeaf
					var = tree{node}.variable;
					if (X(i, var) == 1)
						node = tree{node}.rightchild;
					else
						node = tree{node}.leftchild;
					end
				else
					label = tree{node}.labels;
					if label(1) == label(2)
						pred(i) = randi(2) - 1; % if the counts are ties just randomly assign classes
					else
						pred(i) = label(2) > label(1);
					end
					break;
				end
			end
		end
	end

	function [tree] = buildtree(X, Y, maxdepth)
		global index
		if size(X,2) < maxdepth
		  maxdepth = size(X,2);
		end

		tree = cell(2^size(X,2) + 1, 1);   
		index = 1;
		curnode = 1;
		parent = -1;
		tree = build(tree, parent, curnode, X, Y, maxdepth);
	end

	function [tree] = build(tree, parent, curnode, X, Y, maxdepth)
		global index
		if  index == 1  
		  tree{curnode} = struct('parent', parent, 'depth', 0);
		else
		  tree{curnode} = struct('parent', parent, 'depth', tree{parent}.depth + 1);
		end
		isLeaf = ~(tree{curnode}.depth < maxdepth);
		tree{curnode}.isLeaf = isLeaf;
		if ~isLeaf
			% Choose split var
			[used_vars,~] = varpath(tree, curnode);

			availVars = setdiff(1:size(X,2), used_vars);
			% select a variable to split
			[~, bestIndex] = infogain(X(:,availVars), Y);

			tree{curnode}.variable = availVars(bestIndex);

			% Add child nodes
			leftchild = index + 1;
			rightchild= index + 2;
			index = index + 2;

			tree{curnode}.leftchild = leftchild;
			tree{curnode}.rightchild = rightchild;

			% Depth first search (DFS)
			v = availVars(bestIndex);
			% X(X(:,v)==0,:) is the set of observations split to the left child, and
			% X(X(:,v)==1,:) is is the set of observations split to the right child
			tree = build(tree, curnode, leftchild, X(X(:,v)==0,:), Y(X(:,v)==0,:), maxdepth); % to create a left child with the current node as the parent
			tree = build(tree, curnode, rightchild, X(X(:,v)==1,:), Y(X(:,v)==1,:), maxdepth); % to create a right child with the current node as the parent
		else     
			[y0 y1] = leafcat(tree, curnode, X, Y); % count obs in each class at the leaf nodes
			tree{curnode}.labels = [y0 y1];
		end    
	end
end